### ISSUE TITLE
Using quil with a REPL

### ISSUE THREAD

**Issue by Folcon** [quil/quil#342]
So I've been using quil a bit and one of the problems of using it at the REPL seems to be that every time that I reload the REPL file in REPL it spawns a new window, but if I use `defonce` on the sketch then redefining the draw function for example doesn't change the draw result.

I've taken to using `(var draw)` or `#'draw` as it's shorter and the same with `update` and `setup` for coding in the REPL.

Example:
```
(ns quil-example.core
  (:require [quil.core :as q]
            [quil.middleware :as m]))

(defn setup []
  (q/frame-rate 30)
  (q/color-mode :hsb)
  {:color 140})

(defn update-state [state]
  {:color (mod (+ (:color state) 0.7) 255)})

(defn draw-state [state]
  (q/background (:c state)))

(defn create-sketch []
  (q/sketch
    :title "Quil Example"
    :size [500 500]
    :setup #'setup
    :update #'update-state
    :draw #'draw-state
    :renderer :p3d
    :features [:keep-on-top]
    :middleware [m/fun-mode]))

(defonce sketch (create-sketch))
```
Doing this allows for whole file reload without creating new sketch windows and all of the functions get nicely updated =)...

This seems to be a really nice workflow, but I've not spotted it in the quil docs (perhaps I've missed it). Would it be worth adding something?

There could be a problem with doing this that I'm not aware of so perhaps that's why?

**Comment by nbeloglazov** [comment]
Yes, that's a fair point. Current instructions aim to be as simple as possible with minimal amount of magic even if it leads to a bit clunkier experience like re-spawning window. Your approach requires using `#'` which is a bit of repl-magic that makes the code harder to grasp. But given the benefit it adds - it makes sense to explicitly describe. Can you add a section about your fix [here](https://github.com/quil/quil/wiki/Dynamic-Workflow-(for-REPL))? Somewhere after "(use :reload)".

Also which other docs have found about using quil with repl? There might be different versions laying around.

**Comment by Folcon** [comment]
I've put together a rough bit to start with, I am thinking of writing a longer form guide, but I don't have time at the moment.

We'll see if I get a chance to come back to this =)...

Feel free to edit what I've written if you figure out a more elegant way to put it across =)...

With regards to other guides, I'm not recalling anything at the moment, if I remember I'll add notes here.

**Comment by Xceno** [comment]
Hey guys I might be a bit late to the party, but I just wanted to share my setup here real quick.
I work repl-driven in a `.cljc` file with quil and it works really well for me like this:


```clojure
(defn ^:export new-sketch
  ([host] (new-sketch host 1200 1200))
  ([host width height]
   (q/defsketch floorplan
     :title "My renderer"
     :host host
     :setup setup
     :update update-state
     :draw draw-state
     ;; Ignored in CLJS.
     :features [:keep-on-top :resizable]
     ;; Does nothing, but avoids crashing quil
     ;; due to an internal "bug" in `pause-on-error`
     ;; which tries to call this function with 2 args
     ;; even if it doesn't exist.
     :key-pressed (fn [state _] state)
     :navigation-3d {:rotate-on :mouse-dragged
                              :step-size 200}
     :middleware [#?(:clj qm/pause-on-error)
                  qm/fun-mode
                  qm/navigation-3d]
     :size [width height]
     :renderer :p3d)))
```
What might be of interest is the `:key-pressed` handler in `defsketch`, which is a workaround for the issue @Folcon  mentioned, where you need to restart the sketch when the `pause-on-error` handler blows up.

I also don't need the `var` trick, etc. With this setup everything just works fine.

**Comment by SimplyOnMyWay** [comment]
> Hey guys I might be a bit late to the party, but I just wanted to share my setup here real quick.
> I work repl-driven in a `.cljc` file with quil and it works really well for me like this:
> 
> ```clojure
> (defn ^:export new-sketch
>   ([host] (new-sketch host 1200 1200))
>   ([host width height]
>    (q/defsketch floorplan
>      :title "My renderer"
>      :host host
>      :setup setup
>      :update update-state
>      :draw draw-state
>      ;; Ignored in CLJS.
>      :features [:keep-on-top :resizable]
>      ;; Does nothing, but avoids crashing quil
>      ;; due to an internal "bug" in `pause-on-error`
>      ;; which tries to call this function with 2 args
>      ;; even if it doesn't exist.
>      :key-pressed (fn [state _] state)
>      :navigation-3d {:rotate-on :mouse-dragged
>                               :step-size 200}
>      :middleware [#?(:clj qm/pause-on-error)
>                   qm/fun-mode
>                   qm/navigation-3d]
>      :size [width height]
>      :renderer :p3d)))
> ```
> 
> What might be of interest is the `:key-pressed` handler in `defsketch`, which is a workaround for the issue @Folcon mentioned, where you need to restart the sketch when the `pause-on-error` handler blows up.
> 
> I also don't need the `var` trick, etc. With this setup everything just works fine.

Hi @Xceno!
I have a question different from that of the original poster, but somewhat related.  I'm using clojurescript leaning on the REPL as much as possible to test new functions I write.  I still can't figure out how to use the REPL to call functions from quil.core.  Are you able to do this with your setup?

The REPL throws this when I try...
#object[TypeError TypeError: quil.sketch.current_applet.call(...) is null]

Thanks! 
Micheal

**Comment by UnchartedSystems** [comment]
> Hey guys I might be a bit late to the party, but I just wanted to share my setup here real quick. I work repl-driven in a `.cljc` file with quil and it works really well for me like this:
> 
> ```clojure
> (defn ^:export new-sketch
>   ([host] (new-sketch host 1200 1200))
>   ([host width height]
>    (q/defsketch floorplan
>      :title "My renderer"
>      :host host
>      :setup setup
>      :update update-state
>      :draw draw-state
>      ;; Ignored in CLJS.
>      :features [:keep-on-top :resizable]
>      ;; Does nothing, but avoids crashing quil
>      ;; due to an internal "bug" in `pause-on-error`
>      ;; which tries to call this function with 2 args
>      ;; even if it doesn't exist.
>      :key-pressed (fn [state _] state)
>      :navigation-3d {:rotate-on :mouse-dragged
>                               :step-size 200}
>      :middleware [#?(:clj qm/pause-on-error)
>                   qm/fun-mode
>                   qm/navigation-3d]
>      :size [width height]
>      :renderer :p3d)))
> ```
> 
> What might be of interest is the `:key-pressed` handler in `defsketch`, which is a workaround for the issue @Folcon mentioned, where you need to restart the sketch when the `pause-on-error` handler blows up.
> 
> I also don't need the `var` trick, etc. With this setup everything just works fine.

This key-pressed workaround solves my problems with pause-on-error.
